modules/core/src/resource-management/program-manager.js

```diff
@@ -0,0 +1,120 @@
+import {assembleShaders} from '@luma.gl/shadertools';
+import {Program} from '@luma.gl/webgl';
+
+export default class ProgramManager {
+  constructor(gl) {
+    this.gl = gl;
+
+    this._programCache = {};
+    this._getUniforms = {};
+    this._moduleInjections = {
+      vs: {},
+      fs: {}
+    };
+    this._hookFunctions = {
+      vs: {},
+      fs: {}
+    };
+
+    this._hashes = {};
+    this._hashCounter = 0;
+    this._useCounts = {};
+  }
+
+  addModuleInjection(module, opts) {
+    const moduleName = module.name;
```

Comment:
- What about default shader modules?
- to dedup this and the existing global registration logic, can we create a default program manager and expose its methods for backward compatibility?

Reply:
> What about default shader modules?
What I was imagining is deck could just track what it considers the default shader modules and request programs that include them on each draw, e.g.
```js
// In some layer
draw({uniforms}) {
	this.state.model.setProgram({
       vs,
       fs,
       defaultModules.concat(layerModules)
    });
    this.state.model.draw();
}
```
> to dedup this and the existing global registration logic, can we create a default program manager and expose its methods for backward compatibility?

Is there a way you can think of that wouldn't create a dependency between shadertools and the program manager? I'll add deprecation warning to the global functions in the meantime, since I think they should be removed in 8.

Reply:
We can do whatever under the hood (LayerManager or Layer class), but I feel strongly about not breaking the existing layers. The current standard implementation:

```js
draw({uniforms}) {
	this.state.model.draw({uniforms});
}
```

Should continue to work. Many users write custom layers and we should not add complexity to this process.

Reply:
Sounds good. I think we can use `Layer.drawLayer` and `Layer.getShaders` to get the kind of functionality I'm aiming for. Can we land this implementation as-is, and I'll dig into how things will hook up in deck.